File: scripts/fetch_all.py
#!/usr/bin/env python3
"""
scripts/fetch_all.py
- يجلب /api/all-movies و /api/all-series ويحفظهما كـ JSON.
- Cron-ready. يعتمد على requests.

ENV:
  API_BASE (default: http://localhost:5001)
  OUT_DIR  (default: ./data)

Example:
  API_BASE=http://localhost:5001 OUT_DIR=./data python scripts/fetch_all.py
"""

import os
import time
import json
from datetime import datetime
from pathlib import Path
import requests

API_BASE = os.environ.get("API_BASE", "http://localhost:5001").rstrip("/")
OUT_DIR = Path(os.environ.get("OUT_DIR", "./data"))
RETRIES = 3
RETRY_DELAY = 1.5


def safe_get(url):
    for i in range(RETRIES + 1):
        try:
            r = requests.get(url, timeout=30)
            r.raise_for_status()
            return r.json()
        except Exception as e:
            if i == RETRIES:
                raise
            time.sleep(RETRY_DELAY * (i + 1))


def normalize(item):
    return {
        "title": item.get("title") or item.get("name"),
        "link": item.get("link") or item.get("url"),
        "image": item.get("image"),
        "year": item.get("year"),
        "rating": item.get("rating"),
        "genres": item.get("genres") or [],
        "quality": item.get("quality"),
        "type": item.get("type")
    }


def save_json(name, obj):
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    p = OUT_DIR / name
    p.write_text(json.dumps(obj, indent=2, ensure_ascii=False), encoding="utf-8")
    print("Saved", p)


def main():
    movies_url = f"{API_BASE}/api/all-movies"
    series_url = f"{API_BASE}/api/all-series"
    print("Fetching movies:", movies_url)
    movies_res = safe_get(movies_url)
    print("Fetching series:", series_url)
    series_res = safe_get(series_url)

    movies = [(normalize(i)) for i in (movies_res.get("data") or movies_res)]
    series = [(normalize(i)) for i in (series_res.get("data") or series_res)]

    ts = datetime.utcnow().isoformat().replace(':','-')
    save_json(f"all-movies.{ts}.json", {"fetchedAt": ts, "count": len(movies), "data": movies})
    save_json(f"all-series.{ts}.json", {"fetchedAt": ts, "count": len(series), "data": series})
    save_json("all-movies.latest.json", {"fetchedAt": ts, "count": len(movies), "data": movies})
    save_json("all-series.latest.json", {"fetchedAt": ts, "count": len(series), "data": series})
    print("Done.")


if __name__ == "__main__":
    main()

File: scripts/fetch_all.js
#!/usr/bin/env node
/**
 * scripts/fetch_all.js
 * - Fetches /api/all-movies and /api/all-series and saves JSON files.
 * - Cron-ready. Requires Node 18+ (global fetch). If using older Node, install node-fetch.
 *
 * ENV:
 *   API_BASE (default: http://localhost:5001)
 *   OUT_DIR  (default: ./data)
 *
 * Example:
 *   API_BASE=http://localhost:5001 OUT_DIR=./data node scripts/fetch_all.js
 */

import fs from 'fs/promises';
import path from 'path';

const API_BASE = process.env.API_BASE || 'http://localhost:5001';
const OUT_DIR = process.env.OUT_DIR || './data';
const RETRIES = 3;
const RETRY_DELAY_MS = 1500;

async function sleep(ms){ return new Promise(r => setTimeout(r, ms)); }

async function safeFetch(url, retries = RETRIES){
  for(let i=0;i<=retries;i++){
    try{
      const res = await fetch(url, { timeout: 30000 });
      if(!res.ok) throw new Error(`HTTP ${res.status} ${res.statusText}`);
      const json = await res.json();
      return json;
    }catch(err){
      if(i===retries) throw err;
      await sleep(RETRY_DELAY_MS*(i+1));
    }
  }
}

function normalizeItem(item){
  return {
    title: item.title ?? item.name ?? null,
    link: item.link ?? item.url ?? null,
    image: item.image ?? item.thumbnail ?? null,
    year: item.year ?? item.release_year ?? null,
    rating: item.rating ?? null,
    genres: item.genres ?? item.category ?? [],
    quality: item.quality ?? null,
    type: item.type ?? null
  };
}

async function saveJSON(filename, obj){
  await fs.mkdir(OUT_DIR, { recursive: true });
  const filePath = path.join(OUT_DIR, filename);
  await fs.writeFile(filePath, JSON.stringify(obj, null, 2), 'utf8');
  console.log(`Saved ${filePath}`);
}

async function run(){
  console.log(`API base: ${API_BASE}`);
  try{
    const moviesUrl = `${API_BASE.replace(/\/$/, '')}/api/all-movies`;
    const seriesUrl = `${API_BASE.replace(/\/$/, '')}/api/all-series`;

    console.log('Fetching movies...');
    const moviesRes = await safeFetch(moviesUrl);
    console.log('Fetching series...');
    const seriesRes = await safeFetch(seriesUrl);

    const movies = (moviesRes.data ?? moviesRes.items ?? moviesRes).map(normalizeItem);
    const series = (seriesRes.data ?? seriesRes.items ?? seriesRes).map(normalizeItem);

    const timestamp = new Date().toISOString().replace(/:/g,'-');
    await saveJSON(`all-movies.${timestamp}.json`, { fetchedAt: timestamp, count: movies.length, data: movies });
    await saveJSON(`all-series.${timestamp}.json`, { fetchedAt: timestamp, count: series.length, data: series });

    await saveJSON('all-movies.latest.json', { fetchedAt: timestamp, count: movies.length, data: movies });
    await saveJSON('all-series.latest.json', { fetchedAt: timestamp, count: series.length, data: series });

    console.log('Done.');
  }catch(err){
    console.error('Error:', err.message || err);
    process.exitCode = 1;
  }
}

if (import.meta.url === `file://${process.argv[1]}` || process.argv[1]?.endsWith('fetch_all.js')){
  run();
}
